### *Import Libraries*

In [1]:
import pandas as pd
import numpy as np
import functools
import matplotlib.pyplot as plt
import seaborn as sns
import pandas_profiling
import datetime as dt

### *Load Data*

In [2]:
# Read the csv files for the months Jan, Feb, Mar to dataframe quarter1
quarter1 = pd.concat(map(functools.partial(pd.read_csv, low_memory = False), ['Jan2019.csv', 'Feb2019.csv','Mar2019.csv']))

# Read the csv files for the months Apr, May, Jun to dataframe quarter2
quarter2 = pd.concat(map(functools.partial(pd.read_csv, low_memory = False), ['Apr2019.csv', 'May2019.csv', 'Jun2019.csv']))

# Read the csv files for the months Jul, Aug, Sep to dataframe quarter3
quarter3 = pd.concat(map(functools.partial(pd.read_csv, low_memory = False), ['Jul2019.csv', 'Aug2019.csv','Sep2019.csv']))

# Read the csv files for the months Oct, Nov, Dec to dataframe quarter4
quarter4 = pd.concat(map(functools.partial(pd.read_csv, low_memory = False), ['Oct2019.csv', 'Nov2019.csv', 'Dec2019.csv']))

In [3]:
# Concat the dataframes for all quarters to a dataframe df 
flight_data = pd.concat([quarter1, quarter2, quarter3, quarter4])

### *Inspect Data*

In [4]:
# Call the info() method on flight_data to get the summary of the 
flight_data.shape

(7422037, 78)

In [5]:
# Call the head() method on flight_data to print first five rows
flight_data.head()

YEAR  QUARTER  MONTH  DAY_OF_MONTH  DAY_OF_WEEK     FL_DATE  \
0  2019        1      1             3            4  2019-01-03   
1  2019        1      1             4            5  2019-01-04   
2  2019        1      1             5            6  2019-01-05   
3  2019        1      1             6            7  2019-01-06   
4  2019        1      1             7            1  2019-01-07   

  OP_UNIQUE_CARRIER  OP_CARRIER_AIRLINE_ID OP_CARRIER TAIL_NUM  ...  \
0                9E                  20363         9E   N195PQ  ...   
1                9E                  20363         9E   N919XJ  ...   
2                9E                  20363         9E   N316PQ  ...   
3                9E                  20363         9E   N325PQ  ...   
4                9E                  20363         9E   N904XJ  ...   

   DIV_DISTANCE  DIV1_AIRPORT  DIV1_AIRPORT_ID  DIV1_AIRPORT_SEQ_ID  \
0           NaN           NaN              NaN                  NaN   
1           NaN           NaN              NaN                  NaN   
2           NaN           NaN              NaN                  NaN   
3           NaN           NaN              NaN                  NaN   
4           NaN           NaN              NaN                  NaN   

  DIV1_WHEELS_ON DIV1_TOTAL_GTIME DIV1_LONGEST_GTIME  DIV1_WHEELS_OFF  \
0            NaN              NaN                NaN              NaN   
1            NaN              NaN                NaN              NaN   
2            NaN              NaN                NaN              NaN   
3            NaN              NaN                NaN              NaN   
4            NaN              NaN                NaN              NaN   

  DIV1_TAIL_NUM  Unnamed: 77  
0           NaN          NaN  
1           NaN          NaN  
2           NaN          NaN  
3           NaN          NaN  
4           NaN          NaN  

[5 rows x 78 columns]

In [6]:
# Call the info() method on flight_data to get the summary
flight_data.info(verbose = True, null_counts = True)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7422037 entries, 0 to 625762
Data columns (total 78 columns):
 #   Column                   Non-Null Count    Dtype  
---  ------                   --------------    -----  
 0   YEAR                     7422037 non-null  int64  
 1   QUARTER                  7422037 non-null  int64  
 2   MONTH                    7422037 non-null  int64  
 3   DAY_OF_MONTH             7422037 non-null  int64  
 4   DAY_OF_WEEK              7422037 non-null  int64  
 5   FL_DATE                  7422037 non-null  object 
 6   OP_UNIQUE_CARRIER        7422037 non-null  object 
 7   OP_CARRIER_AIRLINE_ID    7422037 non-null  int64  
 8   OP_CARRIER               7422037 non-null  object 
 9   TAIL_NUM                 7404200 non-null  object 
 10  OP_CARRIER_FL_NUM        7422037 non-null  int64  
 11  ORIGIN_AIRPORT_ID        7422037 non-null  int64  
 12  ORIGIN_AIRPORT_SEQ_ID    7422037 non-null  int64  
 13  ORIGIN_CITY_MARKET_ID    7422037 non-null  

In [7]:
#report = flight_data.profile_report(sort='None', html={'style':{'full_width': True}})
#report

In [8]:
pd.options.display.max_rows = 100

In [9]:
# Call the describe() method on flight_data to get the summary statistics
flight_data.describe().T

count          mean            std        min  \
YEAR                     7422037.0  2.019000e+03       0.000000     2019.0   
QUARTER                  7422037.0  2.525825e+00       1.105914        1.0   
MONTH                    7422037.0  6.578750e+00       3.402585        1.0   
DAY_OF_MONTH             7422037.0  1.573196e+01       8.762183        1.0   
DAY_OF_WEEK              7422037.0  3.937155e+00       1.995814        1.0   
OP_CARRIER_AIRLINE_ID    7422037.0  1.998676e+04     374.751089    19393.0   
OP_CARRIER_FL_NUM        7422037.0  2.557196e+03    1799.413408        1.0   
ORIGIN_AIRPORT_ID        7422037.0  1.264888e+04    1523.847864    10135.0   
ORIGIN_AIRPORT_SEQ_ID    7422037.0  1.264892e+06  152384.558081  1013505.0   
ORIGIN_CITY_MARKET_ID    7422037.0  3.174470e+04    1304.679769    30070.0   
ORIGIN_STATE_FIPS        7422037.0  2.705190e+01      16.456970        1.0   
ORIGIN_WAC               7422037.0  5.403271e+01      26.162471        1.0   
DEST_AIRPORT_ID          7422037.0  1.264881e+04    1523.821061    10135.0   
DEST_AIRPORT_SEQ_ID      7422037.0  1.264885e+06  152381.877700  1013505.0   
DEST_CITY_MARKET_ID      7422037.0  3.174468e+04    1304.645333    30070.0   
DEST_STATE_FIPS          7422037.0  2.705241e+01      16.457194        1.0   
DEST_WAC                 7422037.0  5.403308e+01      26.162656        1.0   
CRS_DEP_TIME             7422037.0  1.330259e+03     492.985466        1.0   
DEP_TIME                 7291951.0  1.334606e+03     507.199818        1.0   
DEP_DELAY                7291927.0  1.092327e+01      48.959640      -82.0   
DEP_DELAY_NEW            7291927.0  1.410478e+01      47.895978        0.0   
DEP_DEL15                7291927.0  1.868044e-01       0.389754        0.0   
DEP_DELAY_GROUP          7291927.0  8.095898e-02       2.276868       -2.0   
TAXI_OUT                 7288060.0  1.738935e+01      10.003982        1.0   
WHEELS_OFF               7288060.0  1.358264e+03     508.810907        1.0   
WHEELS_ON                7284390.0  1.459032e+03     537.841369        1.0   
TAXI_IN                  7284390.0  7.740776e+00       6.194340        1.0   
CRS_ARR_TIME             7422037.0  1.485790e+03     521.465892        1.0   
ARR_TIME                 7284391.0  1.462955e+03     542.454249        1.0   
ARR_DELAY                7268232.0  5.414849e+00      51.066656      -99.0   
ARR_DELAY_NEW            7268232.0  1.416780e+01      47.645489        0.0   
ARR_DEL15                7268232.0  1.911404e-01       0.393199        0.0   
ARR_DELAY_GROUP          7268232.0 -1.838245e-01       2.422859       -2.0   
CANCELLED                7422037.0  1.817897e-02       0.133598        0.0   
DIVERTED                 7422037.0  2.543776e-03       0.050372        0.0   
CRS_ELAPSED_TIME         7421902.0  1.419022e+02      72.346849        1.0   
ACTUAL_ELAPSED_TIME      7268232.0  1.366903e+02      72.612742       15.0   
AIR_TIME                 7268232.0  1.115734e+02      70.561226        4.0   
FLIGHTS                  7422037.0  1.000000e+00       0.000000        1.0   
DISTANCE                 7422037.0  8.005437e+02     592.511766       31.0   
DISTANCE_GROUP           7422037.0  3.676005e+00       2.332646        1.0   
CARRIER_DELAY            1389253.0  2.112859e+01      66.104645        0.0   
WEATHER_DELAY            1389253.0  3.802404e+00      32.362973        0.0   
NAS_DELAY                1389253.0  1.658795e+01      39.664395        0.0   
SECURITY_DELAY           1389253.0  9.608329e-02       3.358346        0.0   
LATE_AIRCRAFT_DELAY      1389253.0  2.740748e+01      53.456198        0.0   
FIRST_DEP_TIME             51136.0  1.343648e+03     506.654217        1.0   
TOTAL_ADD_GTIME            51135.0  4.023841e+01      35.555019        1.0   
LONGEST_ADD_GTIME          51135.0  3.961257e+01      34.349323        1.0   
DIV_AIRPORT_LANDINGS     7422037.0  3.711919e-03       0.113724        0.0   
DIV_REACHED_DEST           18880.0  8.558792e-01   

### *Missing Values*

In [10]:
# Call isnull() and mean() on flight_data to get the percentage of missing values for each column
round(flight_data.isnull().mean()[flight_data.isnull().mean().gt(0.75)] * 100, 2)

CANCELLATION_CODE           98.18
CARRIER_DELAY               81.28
WEATHER_DELAY               81.28
NAS_DELAY                   81.28
SECURITY_DELAY              81.28
LATE_AIRCRAFT_DELAY         81.28
FIRST_DEP_TIME              99.31
TOTAL_ADD_GTIME             99.31
LONGEST_ADD_GTIME           99.31
DIV_REACHED_DEST            99.75
DIV_ACTUAL_ELAPSED_TIME     99.78
DIV_ARR_DELAY               99.78
DIV_DISTANCE                99.75
DIV1_AIRPORT                99.73
DIV1_AIRPORT_ID             99.73
DIV1_AIRPORT_SEQ_ID         99.73
DIV1_WHEELS_ON              99.73
DIV1_TOTAL_GTIME            99.73
DIV1_LONGEST_GTIME          99.73
DIV1_WHEELS_OFF             99.78
DIV1_TAIL_NUM               99.78
Unnamed: 77                100.00
dtype: float64

In [11]:
# Extract the columns with percentage of null values greater than 75% to a dataframe null_cols
null_cols = (flight_data.isnull().mean()[flight_data.isnull().mean().gt(0.75)] * 100).to_frame()

# Extract the column names to be dropped
cols_to_drop = list(null_cols.index)

In [12]:
# Call drop() method on flight data 
# to drop columns with more than 75% of null values
flight_data = flight_data.drop(cols_to_drop, axis = 1)

In [13]:
# Check for null values
flight_data.isnull().sum()[flight_data.isnull().sum().gt(0)]

TAIL_NUM                17837
DEP_TIME               130086
DEP_DELAY              130110
DEP_DELAY_NEW          130110
DEP_DEL15              130110
DEP_DELAY_GROUP        130110
TAXI_OUT               133977
WHEELS_OFF             133977
WHEELS_ON              137647
TAXI_IN                137647
ARR_TIME               137646
ARR_DELAY              153805
ARR_DELAY_NEW          153805
ARR_DEL15              153805
ARR_DELAY_GROUP        153805
CRS_ELAPSED_TIME          135
ACTUAL_ELAPSED_TIME    153805
AIR_TIME               153805
dtype: int64

*'CRS_ELAPSED_TIME' - "Estmated time an airplane spends in the air, as opposed to time spent taxiing to and from the gate and during stopovers.*  

*Has 135 missing values that could be imputed or dropped..since the null value count is too small when comapred to the dataset we have, we can drop the nulls*

In [14]:
# Drop the rows with null values for column CRS_ELAPSED_TIME 
flight_data.dropna(subset = ['CRS_ELAPSED_TIME'], inplace = True)

In [15]:
# Get the shape of flight_data
flight_data.shape

(7421902, 56)

### *Duplicates*

In [16]:
# Find the duplicate rows
duplicates = flight_data.duplicated()

In [17]:
flight_data[duplicates]

Empty DataFrame
Columns: [YEAR, QUARTER, MONTH, DAY_OF_MONTH, DAY_OF_WEEK, FL_DATE, OP_UNIQUE_CARRIER, OP_CARRIER_AIRLINE_ID, OP_CARRIER, TAIL_NUM, OP_CARRIER_FL_NUM, ORIGIN_AIRPORT_ID, ORIGIN_AIRPORT_SEQ_ID, ORIGIN_CITY_MARKET_ID, ORIGIN, ORIGIN_CITY_NAME, ORIGIN_STATE_ABR, ORIGIN_STATE_FIPS, ORIGIN_STATE_NM, ORIGIN_WAC, DEST_AIRPORT_ID, DEST_AIRPORT_SEQ_ID, DEST_CITY_MARKET_ID, DEST, DEST_CITY_NAME, DEST_STATE_ABR, DEST_STATE_FIPS, DEST_STATE_NM, DEST_WAC, CRS_DEP_TIME, DEP_TIME, DEP_DELAY, DEP_DELAY_NEW, DEP_DEL15, DEP_DELAY_GROUP, DEP_TIME_BLK, TAXI_OUT, WHEELS_OFF, WHEELS_ON, TAXI_IN, CRS_ARR_TIME, ARR_TIME, ARR_DELAY, ARR_DELAY_NEW, ARR_DEL15, ARR_DELAY_GROUP, ARR_TIME_BLK, CANCELLED, DIVERTED, CRS_ELAPSED_TIME, ACTUAL_ELAPSED_TIME, AIR_TIME, FLIGHTS, DISTANCE, DISTANCE_GROUP, DIV_AIRPORT_LANDINGS]
Index: []

[0 rows x 56 columns]

### *DateTime Object*

In [18]:
# Convert 'FL_DATE' column from string to datetime
flight_data['FL_DATE'] = pd.to_datetime(flight_data['FL_DATE']) #.dt.date

In [19]:
flight_data.head().T

0                    1  \
YEAR                                  2019                 2019   
QUARTER                                  1                    1   
MONTH                                    1                    1   
DAY_OF_MONTH                             3                    4   
DAY_OF_WEEK                              4                    5   
FL_DATE                2019-01-03 00:00:00  2019-01-04 00:00:00   
OP_UNIQUE_CARRIER                       9E                   9E   
OP_CARRIER_AIRLINE_ID                20363                20363   
OP_CARRIER                              9E                   9E   
TAIL_NUM                            N195PQ               N919XJ   
OP_CARRIER_FL_NUM                     5121                 5121   
ORIGIN_AIRPORT_ID                    15412                15412   
ORIGIN_AIRPORT_SEQ_ID              1541205              1541205   
ORIGIN_CITY_MARKET_ID                35412                35412   
ORIGIN                                 TYS                  TYS   
ORIGIN_CITY_NAME             Knoxville, TN        Knoxville, TN   
ORIGIN_STATE_ABR                        TN                   TN   
ORIGIN_STATE_FIPS                       47                   47   
ORIGIN_STATE_NM                  Tennessee            Tennessee   
ORIGIN_WAC                              54                   54   
DEST_AIRPORT_ID                      10397                10397   
DEST_AIRPORT_SEQ_ID                1039707              1039707   
DEST_CITY_MARKET_ID                  30397                30397   
DEST                                   ATL                  ATL   
DEST_CITY_NAME                 Atlanta, GA          Atlanta, GA   
DEST_STATE_ABR                          GA                   GA   
DEST_STATE_FIPS                         13                   13   
DEST_STATE_NM                      Georgia              Georgia   
DEST_WAC                                34                   34   
CRS_DEP_TIME                          1140                 1140   
DEP_TIME                              1205                 1250   
DEP_DELAY                               25                   70   
DEP_DELAY_NEW                           25                   70   
DEP_DEL15                                1                    1   
DEP_DELAY_GROUP                          1                    4   
DEP_TIME_BLK                     1100-1159            1100-1159   
TAXI_OUT                                30                   35   
WHEELS_OFF                            1235                 1325   
WHEELS_ON                             1311                 1403   
TAXI_IN                                  4                    9   
CRS_ARR_TIME                          1250                 1250   
ARR_TIME                              1315                 1412   
ARR_DELAY                               25                   82   
ARR_DELAY_NEW                           25                   82   
ARR_DEL15                                1                    1   
ARR_DELAY_GROUP                          1                    5   
ARR_TIME_BLK                     1200-1259            1200-1259   
CANCELLED                                0                    0   
DIVERTED                                 0                    0   
CRS_ELAPSED_TIME                        70                   70   
ACTUAL_ELAPSED_TIME                     70                   82   
AIR_TIME                                36                   38   
FLIGHTS                                  1                    1   
DISTANCE                               152                  152   
DISTANCE_GROUP                           1                    1   
DIV_AIRPORT_LANDINGS                     0                    0   

                                         2                    3  \
YEAR                                  2019                 2019   
QUARTER                                  1                    1   
MONTH               

### *Save clean datafarame to csv*

In [20]:
#Save clean dataframe to csv
flight_data.to_csv('flight_data_cleaned.csv', index = False)